### PR TITLE
I457 -  API returns list of available demographic stat types for a touchstone

### PR DESCRIPTION
**Pull request by littledice**

### CODE REVIEW

**Comment by littledice** [comment]
Also requires: https://github.com/vimc/montagu-webmodels/pull/6

File: spec/spec.md [hunk 1]
```diff
@@ -1698,13 +1698,19 @@ Schema: [`Demographics.schema.json`](Demographics.schema.json)
 
 ### Example
     [{ 
-        "code" : "as-fert",
+        "id" : "as-fert",
         "name": "Age-specific fertility",
+        "variants" : ["low", "medium"],
```
Comment (littledice) [inline spec/spec.md]: maybe we should also return default variant here?

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt [hunk 2]
```diff
@@ -122,6 +167,19 @@ class JooqTouchstoneRepository(
             mapEnum(record.status)
     )
 
+    fun mapDemographicStatisticType(records: List<Record>): DemographicStatisticType
+    {
+        val record = records[0]
```
Comment (MartinEden) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt]: Will this always be one row? If so, better to use `single`. If there could be more rows, could you add a comment explaining why it's fine to just use the first one? Thanks.

File: spec/spec.md (hunk 1 above)
Comment (littledice) [inline spec/spec.md]: hmm I thought I took this out. I think we remove all references to variants for now

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            )})
+        }
+    }
+
+    private var _countries: List<String> = listOf()
+    private var _sourceIds: List<Int> = listOf()
+    private var _sources: List<String> = listOf("UNWPP2015", "UNWPP2017")
+    private var _variantIds: List<Int> = listOf()
+    private var _variants = listOf("low", "medium", "high")
+    private var _units: List<Int> = listOf()
```
Comment (MartinEden) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt]: I'd prefer we didn't use underscores to denote private fields - I'm generally not a big fan of Hungarian notation.

File: spec/DemographicDataset.schema.json [hunk 4]
```diff
@@ -2,10 +2,13 @@
     "id": "DemographicDataSet",
     "type": "object",
     "properties": {
-        "code": { "$ref": "StandardIdentifier.schema.json" },
+        "id": { "$ref": "StandardIdentifier.schema.json" },
         "name": { "type": "string" },
+        "source": { "type": "string" },
+        "countries": { "type": "array", "items": { "$ref": "CountryId.schema.json"  } },
+        "variants": { "type": "array", "items": { "type": "string" } },
```
Comment (littledice) [inline spec/DemographicDataset.schema.json]: this is weird - I actually removed variants from the spec, I don't know why the diff is still showing it

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt (hunk 3 above)
Comment (MartinEden) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt]: Also, a more Kotlin-y way might be this. I mean, what you have is totally fine, I was just wondering what the 'var-less' approach would be...

```
private data class SupportingTables(db: JooqContext)
{
    val countries = db.generateCountries(3)
    val sources = listOf("UNWPP2015", "UNWPP2017")
    val sourceIds = db.generateDemographicSources(sources)
    val variants = listOf("low", "medium", "high")
    val variantIds = it.generateDemographicVariants(variants)
    val units = it.generateDemographicUnits()

    init
    {
        it.generateGenders()
    }
}
```

And then:
```
validate("/touchstones/$touchstoneId/demographics/") against "Demographics" given {
            val supportingTables = SupportingTables(it)
            setUpTouchstone(it, supportingTables)
            addPopulation(it, supportingTables)
}
...
```

File: src/databaseTests/src/test/kotlin/org/vaccineimpact/api/databaseTests/touchstoneRepository/GetDemographicsTests.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    private var _variants = listOf("low", "medium", "high")
+    private var _units: List<Int> = listOf()
+
+    private fun setUpSupportingTables(it: JooqContext)
+    {
+        _countries = it.generateCountries(3)
+        _sourceIds = it.generateDemographicSources(_sources)
+        _variantIds = it.generateDemographicVariants(_variants)
+        _units = it.generateDemographicUnits()
+        it.generateGenders()
```
Comment (MartinEden) [inline src/databaseTests/src/test/kotlin/org/vaccineimpact/api/databaseTests/touchstoneRepository/GetDemographicsTests.kt]: Same thing here (Hungarian notation).

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt (hunk 3 above)
Comment (littledice) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt]: Cool, I'll leave these as is but next time try this approach!

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt [hunk 6]
```diff
@@ -7,7 +7,7 @@ import org.vaccineimpact.api.blackboxTests.helpers.validate
 import org.vaccineimpact.api.db.JooqContext
 import org.vaccineimpact.api.db.direct.*
 import org.vaccineimpact.api.models.permissions.PermissionSet
-import org.vaccineimpact.api.test_helpers.DatabaseTest
+import org.vaccineimpact.api.testhelpers.DatabaseTest
```
Comment (MartinEden) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt]: Whoops - overzealous find + replace

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt [hunk 7]
```diff
@@ -32,39 +32,39 @@ class DemographicTests: DatabaseTest()
                     "id" to "tot-pop",
                     "name" to "tot-pop descriptive name",
                     "source" to "UNWPP2015 descriptive name",
-                    "countries" to array(_countries.sortedBy { it }),
-                    "gender_is_applicable" to false
+                    "countries" to array(countries.sortedBy { it }),
+                    "genderisapplicable" to false
```
Comment (MartinEden) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/DemographicTests.kt]: Overzealous f+r

**Pull request merged**